core/src/main/java/org/apache/calcite/util/BuiltInMethod.java

```diff
@@ -417,6 +417,7 @@ public enum BuiltInMethod {
   INITCAP(SqlFunctions.class, "initcap", String.class),
   SUBSTRING(SqlFunctions.class, "substring", String.class, int.class,
       int.class),
+  FORMAT_NUMBER(SqlFunctions.class, "formatNumber", int.class, int.class),
```

Comment:
How do the functions with `(int, string)` signature get called when you specify `(int.class, int.class)` here? Not sure if it causes problems, just curious.

Reply:
The declared parameters type is only used to check whether there is a matched implemented method in the declared class, the related code line is [link](https://github.com/apache/calcite/blob/b2cdab28ad154f20dfb5ce19ec7fd4d772d6ea2c/core/src/main/java/org/apache/calcite/util/BuiltInMethod.java#L766). It doesn't check all implemented methods, it just validates only one, and if not found, it throws `NoSuchMethodException`. So this is a weak check here.

IMO, there are two phases to validate function's parameters type and arguments type:

In validate phase, the function's parameters type will be validated.

In execution phase, the generated code for expression will be compiled to class, if there is no matched method, including not matched parameters type, it will throws compilation error. Notice the code generation phase doesn't check the matchness of actual arguments type and function's parameters type.

Please point out if there is something wrong.